{
  "cve_id": "CVE-2022-35994",
  "cve_desc": "TensorFlow is an open source platform for machine learning. When `CollectiveGather` receives an scalar input `input`, it gives a `CHECK` fails that can be used to trigger a denial of service attack. We have patched the issue in GitHub commit c1f491817dec39a26be3c574e86a88c30f3c4770. The fix will be included in TensorFlow 2.10.0. We will also cherrypick this commit on TensorFlow 2.9.1, TensorFlow 2.8.1, and TensorFlow 2.7.2, as these are also affected and still in supported range. There are no known workarounds for this issue.",
  "repo": "tensorflow/tensorflow",
  "patch_hash": "c1f491817dec39a26be3c574e86a88c30f3c4770",
  "patch_info": {
    "commit_hash": "c1f491817dec39a26be3c574e86a88c30f3c4770",
    "repo": "tensorflow/tensorflow",
    "commit_url": "https://github.com/tensorflow/tensorflow/commit/c1f491817dec39a26be3c574e86a88c30f3c4770",
    "files": [
      "tensorflow/core/kernels/collective_ops.cc",
      "tensorflow/python/ops/collective_ops_test.py"
    ],
    "message": "Fix for check failure in CollectiveGather op.\n\nThe fix is required only for Eager mode, graph mode already checked for input shape in shape inference pass.\n\nPiperOrigin-RevId: 460801136",
    "before_after_code_files": [
      "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc",
      "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py"
    ]
  },
  "patch_diff": {
    "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc": [
      "File: tensorflow/core/kernels/collective_ops.cc -> tensorflow/core/kernels/collective_ops.cc",
      "--- Hunk 1 ---",
      "[Context before]",
      "176:   void ComputeAsyncImpl(OpKernelContext* c, CollectiveExecutor* col_exec,",
      "177:                         DoneCallback done) override {",
      "178:     auto output_shape = c->input(0).shape();",
      "179:     output_shape.set_dim(",
      "180:         0, output_shape.dim_size(0) * col_params_->group.group_size);",
      "181:     col_params_->instance.shape = output_shape;",
      "",
      "[Removed Lines]",
      "[None]",
      "",
      "[Added Lines]",
      "179:     OP_REQUIRES_ASYNC(c, output_shape.dims() > 0,",
      "180:                       errors::InvalidArgument(\"input should have rank > 0, \",",
      "181:                                               \"recieved \", output_shape.dims()),",
      "182:                       done);",
      "",
      "---------------"
    ],
    "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py": [
      "File: tensorflow/python/ops/collective_ops_test.py -> tensorflow/python/ops/collective_ops_test.py",
      "--- Hunk 1 ---",
      "[Context before]",
      "451:     ])",
      "452:     context.ensure_initialized()",
      "454:     @def_function.function",
      "455:     def run_all_reduce():",
      "456:       group_key = 10",
      "",
      "[Removed Lines]",
      "[None]",
      "",
      "[Added Lines]",
      "454:   @test_util.run_v2_only",
      "455:   def testCollectiveGatherShapeCheckFailure(self):",
      "456:     with self.assertRaisesRegex(errors.InvalidArgumentError,",
      "457:                                 'input should have rank > 0'):",
      "458:       collective_ops.gen_collective_ops.CollectiveGather(",
      "459:           input=1,",
      "460:           group_size=1,",
      "461:           group_key=1,",
      "462:           instance_key=1,",
      "463:           shape=(3, 3, 3),",
      "464:           communication_hint='auto',",
      "465:           timeout_seconds=0,",
      "466:           name='')",
      "",
      "---------------"
    ]
  },
  "candidates": [
    {
      "candidate_hash": "b709df75d9dc61bf53f34f6c509de15e99c0aba4",
      "candidate_info": {
        "commit_hash": "b709df75d9dc61bf53f34f6c509de15e99c0aba4",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/b709df75d9dc61bf53f34f6c509de15e99c0aba4",
        "files": [
          "tensorflow/core/kernels/collective_ops.cc",
          "tensorflow/python/ops/collective_ops_test.py"
        ],
        "message": "Fix for check failure in CollectiveGather op.\n\nThe fix is required only for Eager mode, graph mode already checked for input shape in shape inference pass.\n\nPiperOrigin-RevId: 460801136",
        "before_after_code_files": [
          "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc",
          "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc",
            "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py"
          ],
          "candidate": [
            "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc",
            "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc": [
          "File: tensorflow/core/kernels/collective_ops.cc -> tensorflow/core/kernels/collective_ops.cc",
          "--- Hunk 1 ---",
          "[Context before]",
          "176:   void ComputeAsyncImpl(OpKernelContext* c, CollectiveExecutor* col_exec,",
          "177:                         DoneCallback done) override {",
          "178:     auto output_shape = c->input(0).shape();",
          "179:     output_shape.set_dim(",
          "180:         0, output_shape.dim_size(0) * col_params_->group.group_size);",
          "181:     col_params_->instance.shape = output_shape;",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "179:     OP_REQUIRES_ASYNC(c, output_shape.dims() > 0,",
          "180:                       errors::InvalidArgument(\"input should have rank > 0, \",",
          "181:                                               \"recieved \", output_shape.dims()),",
          "182:                       done);",
          "",
          "---------------"
        ],
        "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py": [
          "File: tensorflow/python/ops/collective_ops_test.py -> tensorflow/python/ops/collective_ops_test.py",
          "--- Hunk 1 ---",
          "[Context before]",
          "451:     ])",
          "452:     context.ensure_initialized()",
          "454:     @def_function.function",
          "455:     def run_all_reduce():",
          "456:       group_key = 10",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "454:   @test_util.run_v2_only",
          "455:   def testCollectiveGatherShapeCheckFailure(self):",
          "456:     with self.assertRaisesRegex(errors.InvalidArgumentError,",
          "457:                                 'input should have rank > 0'):",
          "458:       collective_ops.gen_collective_ops.CollectiveGather(",
          "459:           input=1,",
          "460:           group_size=1,",
          "461:           group_key=1,",
          "462:           instance_key=1,",
          "463:           shape=(3, 3, 3),",
          "464:           communication_hint='auto',",
          "465:           timeout_seconds=0,",
          "466:           name='')",
          "",
          "---------------"
        ]
      }
    },
    {
      "candidate_hash": "01cf6e2a26f01deac681248d02c7c12706519374",
      "candidate_info": {
        "commit_hash": "01cf6e2a26f01deac681248d02c7c12706519374",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/01cf6e2a26f01deac681248d02c7c12706519374",
        "files": [
          "tensorflow/core/kernels/collective_ops.cc",
          "tensorflow/python/ops/collective_ops_test.py"
        ],
        "message": "Fix for check failure in CollectiveGather op.\n\nThe fix is required only for Eager mode, graph mode already checked for input shape in shape inference pass.\n\nPiperOrigin-RevId: 460801136",
        "before_after_code_files": [
          "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc",
          "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc",
            "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py"
          ],
          "candidate": [
            "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc",
            "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc": [
          "File: tensorflow/core/kernels/collective_ops.cc -> tensorflow/core/kernels/collective_ops.cc",
          "--- Hunk 1 ---",
          "[Context before]",
          "176:   void ComputeAsyncImpl(OpKernelContext* c, CollectiveExecutor* col_exec,",
          "177:                         DoneCallback done) override {",
          "178:     auto output_shape = c->input(0).shape();",
          "179:     output_shape.set_dim(",
          "180:         0, output_shape.dim_size(0) * col_params_->group.group_size);",
          "181:     col_params_->instance.shape = output_shape;",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "179:     OP_REQUIRES_ASYNC(c, output_shape.dims() > 0,",
          "180:                       errors::InvalidArgument(\"input should have rank > 0, \",",
          "181:                                               \"recieved \", output_shape.dims()),",
          "182:                       done);",
          "",
          "---------------"
        ],
        "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py": [
          "File: tensorflow/python/ops/collective_ops_test.py -> tensorflow/python/ops/collective_ops_test.py",
          "--- Hunk 1 ---",
          "[Context before]",
          "455:     ])",
          "456:     context.ensure_initialized()",
          "458:     @def_function.function",
          "459:     def run_all_reduce():",
          "460:       group_key = 10",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "458:   @test_util.run_v2_only",
          "459:   def testCollectiveGatherShapeCheckFailure(self):",
          "460:     with self.assertRaisesRegex(errors.InvalidArgumentError,",
          "461:                                 'input should have rank > 0'):",
          "462:       collective_ops.gen_collective_ops.CollectiveGather(",
          "463:           input=1,",
          "464:           group_size=1,",
          "465:           group_key=1,",
          "466:           instance_key=1,",
          "467:           shape=(3, 3, 3),",
          "468:           communication_hint='auto',",
          "469:           timeout_seconds=0,",
          "470:           name='')",
          "",
          "---------------"
        ]
      }
    },
    {
      "candidate_hash": "efddbd01056197ece20c57800324b75c227b1cf1",
      "candidate_info": {
        "commit_hash": "efddbd01056197ece20c57800324b75c227b1cf1",
        "repo": "tensorflow/tensorflow",
        "commit_url": "https://github.com/tensorflow/tensorflow/commit/efddbd01056197ece20c57800324b75c227b1cf1",
        "files": [
          "tensorflow/core/kernels/collective_ops.cc",
          "tensorflow/python/ops/collective_ops_test.py"
        ],
        "message": "Fix for check failure in CollectiveGather op.\n\nThe fix is required only for Eager mode, graph mode already checked for input shape in shape inference pass.\n\nPiperOrigin-RevId: 460801136",
        "before_after_code_files": [
          "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc",
          "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py"
        ]
      },
      "candidate_patch_features": {
        "candidate_earlier_than_patch": 0,
        "diff_branch_same_aad": 1,
        "olp_code_files": {
          "patch": [
            "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc",
            "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py"
          ],
          "candidate": [
            "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc",
            "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py"
          ]
        }
      },
      "candidate_diff": {
        "tensorflow/core/kernels/collective_ops.cc||tensorflow/core/kernels/collective_ops.cc": [
          "File: tensorflow/core/kernels/collective_ops.cc -> tensorflow/core/kernels/collective_ops.cc",
          "--- Hunk 1 ---",
          "[Context before]",
          "176:   void ComputeAsyncImpl(OpKernelContext* c, CollectiveExecutor* col_exec,",
          "177:                         DoneCallback done) override {",
          "178:     auto output_shape = c->input(0).shape();",
          "179:     output_shape.set_dim(",
          "180:         0, output_shape.dim_size(0) * col_params_->group.group_size);",
          "181:     col_params_->instance.shape = output_shape;",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "179:     OP_REQUIRES_ASYNC(c, output_shape.dims() > 0,",
          "180:                       errors::InvalidArgument(\"input should have rank > 0, \",",
          "181:                                               \"recieved \", output_shape.dims()),",
          "182:                       done);",
          "",
          "---------------"
        ],
        "tensorflow/python/ops/collective_ops_test.py||tensorflow/python/ops/collective_ops_test.py": [
          "File: tensorflow/python/ops/collective_ops_test.py -> tensorflow/python/ops/collective_ops_test.py",
          "--- Hunk 1 ---",
          "[Context before]",
          "451:     ])",
          "452:     context.ensure_initialized()",
          "454:     @def_function.function",
          "455:     def run_all_reduce():",
          "456:       group_key = 10",
          "",
          "[Removed Lines]",
          "[None]",
          "",
          "[Added Lines]",
          "454:   @test_util.run_v2_only",
          "455:   def testCollectiveGatherShapeCheckFailure(self):",
          "456:     with self.assertRaisesRegex(errors.InvalidArgumentError,",
          "457:                                 'input should have rank > 0'):",
          "458:       collective_ops.gen_collective_ops.CollectiveGather(",
          "459:           input=1,",
          "460:           group_size=1,",
          "461:           group_key=1,",
          "462:           instance_key=1,",
          "463:           shape=(3, 3, 3),",
          "464:           communication_hint='auto',",
          "465:           timeout_seconds=0,",
          "466:           name='')",
          "",
          "---------------"
        ]
      }
    }
  ]
}